src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,143 +112,226 @@ public class FromBase64Transform : ICryptoTransform
+
+            // RyuJIT produces better code with the ternary. This JIT-bug is tracked in https://github.com/dotnet/coreclr/issues/914
+            return value == 32 || ((uint)value - 9 <= (13 - 9));
         }
 
-        private byte[] DiscardWhiteSpaces(byte[] inputBuffer, int inputOffset, int inputCount)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetOutputSize(int bytesToTransform, Span<byte> tmpBuffer)
         {
-            int i, iCount = 0;
-            for (i = 0; i < inputCount; i++)
+            int outputSize = Base64.GetMaxDecodedFromUtf8Length(bytesToTransform);
+
+            const byte padding = (byte)'=';
+            int len = tmpBuffer.Length;
+
+            // In Base64 there are maximum 2 padding chars
+
+            if (tmpBuffer[len - 2] == padding)
             {
-                if (char.IsWhiteSpace((char)inputBuffer[inputOffset + i])) iCount++;
+                outputSize--;
             }
 
-            // If there's nothing to do, leave early
-            if (iCount == 0 && inputOffset == 0 &&
-                inputCount == inputBuffer.Length)
+            if (tmpBuffer[len - 1] == padding)
             {
-                return inputBuffer;
+                outputSize--;
             }
 
-            byte[] rgbOut = new byte[inputCount - iCount];
-            iCount = 0;
-            for (i = 0; i < inputCount; i++)
+            return outputSize;
+        }
+
+        private void ConvertFromBase64(Span<byte> tmpBuffer, Span<byte> outputBuffer, out int consumed, out int written)
+        {
+            int bytesToTransform = _inputIndex + tmpBuffer.Length;
+
+            // Common case for bytesToTransform = 4
+            byte[] transformBufferArray = null;
+            Span<byte> transformBuffer = stackalloc byte[4];
+            if (bytesToTransform > 4)
             {
-                if (!char.IsWhiteSpace((char)inputBuffer[inputOffset + i]))
-                {
-                    rgbOut[iCount++] = inputBuffer[inputOffset + i];
-                }
+                transformBuffer = transformBufferArray = CryptoPool.Rent(bytesToTransform);
             }
 
-            return rgbOut;
+            transformBuffer = transformBuffer.Slice(0, bytesToTransform);
+
+            // Copy _inputBuffer to transformBuffer and append tmpBuffer
+            Debug.Assert(_inputIndex < _inputBuffer.Length);
+            _inputBuffer.AsSpan(0, _inputIndex).CopyTo(transformBuffer);
+            tmpBuffer.CopyTo(transformBuffer.Slice(_inputIndex));
+
+            // Save data that won't be transformed to _inputBuffer, so it can be transformed later
+            _inputIndex = bytesToTransform & 3;     // bit hack for % 4
+            Debug.Assert(_inputIndex < _inputBuffer.Length);
+            tmpBuffer.Slice(tmpBuffer.Length - _inputIndex).CopyTo(_inputBuffer);
+
+            OperationStatus status = Base64.DecodeFromUtf8(transformBuffer, outputBuffer, out consumed, out written);
```

Comment:
Here is a bug, depending on the value of `bytesToTransform`, that is not covered with the current tests.
With the current implementation `status` may becode `InvalidData`, which is not correct.
The correct behavior should only try to transform up to the next multiple of 4 (the rest is stored to `_inputBuffer` for later use).

I've tests therefore and a fix. Will push them tomorrow.

Reply:
Tests and fix --> ad6ca8b59b7004b44f5cb1d526006ccd83c19b68